### PR TITLE
Functional tests for interventions

### PR DESCRIPTION
**Pull request by jiayuanz3**
## Summary

Describe the issue, and what you have done to solve it

## Further work

If relevant, add in what needs to be done next on this area (and whether you will do it!)

## Checklist

- [ ] All new functions have docstrings in the correct style
- [ ] I've verified the complete docs build locally without errors
- [ ] I've maintained 100% coverage (please mention any 'no cover' annotations explicitly)
- [ ] I've unit-tested all new methods directly

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have read the **CONTRIBUTING** document.
- [ ] My code follows the code style of this project.
- [ ] I have updated the wiki with new parameters/model functionality

## Closing issues

Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: pyEpiabm/pyEpiabm/tests/test_func/test_distancing_functional.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            'distancing_enhanced_prob'] = [0.1]*17
+        pop = TestDistancingFunctional.file_simulation(
+            "test_input.csv", self.sim_params, self.file_params,
+            HelperFunc().sweep_list_initialise())
+
+        mock_read.assert_called_with('test_input.csv')
+        self.assertEqual(mock_csv.call_count, 2)
+
+        # Compare number of susceptible individuals for each age group
+        HelperFunc().assert_greater_equal(
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/tests/test_func/test_distancing_functional.py:198]: Have you thought about the possibility of tests like this failing by chance? Is there any reason why this has to be the case, or just that it is most likely that these age groups with lower social distancing will have more infections?

**Review by KCGallagher — COMMENTED** [review]
Looks good! See comments - because of the amount of copied code in this PR, many of them apply in multiple places so make sure you change wherever appropriate

File: pyEpiabm/pyEpiabm/tests/test_func/helper_func.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import pyEpiabm as pe
+from pyEpiabm.property.infection_status import InfectionStatus
+
+
+class HelperFunc(unittest.TestCase):
+    """Class to help running functional tests for interventions.
+
+    """
+
+    def assert_greater_equal(self, small_pop, large_pop,
```
[the remark this answers is not in the document]
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_func/helper_func.py:16]: Done!

File: pyEpiabm/pyEpiabm/tests/test_func/helper_func.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        method: string
+            Specify if the comparing is for greater or equal or only equal
+
+        """
+        for age_group in range(len(pe.Parameters.instance().age_proportions)):
+            with self.subTest(age_group=age_group):
+                if method == 'greater':
+                    self.assertGreaterEqual(
+                        large_pop[0].compartment_counter.retrieve()[
+                            InfectionStatus.Susceptible][age_group],
```
[the remark this answers is not in the document]
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_func/helper_func.py:39]: Sorry, I forgot to change this yesterday.

File: pyEpiabm/pyEpiabm/tests/test_func/helper_func.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_func/helper_func.py:16]: But as I'm using unittest for testing, I have to inherit unittest.TestCase in the begin of the class. So I think I still have to keep it ?

File: pyEpiabm/pyEpiabm/tests/test_func/test_closure_functional.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
+import os
+import pandas as pd
+import unittest
+from unittest.mock import patch, mock_open, Mock
+
+import pyEpiabm as pe
+from helper_func import HelperFunc
```
[the remark this answers is not in the document]
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_func/test_closure_functional.py:7]: Done!

File: pyEpiabm/pyEpiabm/tests/test_func/test_closure_functional.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        # Compare number of susceptible individuals for each age group
+        HelperFunc().assert_greater_equal(
+             pop.cells, pop_closure.cells)
+
+    @patch('pyEpiabm.routine.simulation.tqdm', notqdm)
+    @patch('pyEpiabm.output._CsvDictWriter.write', Mock())
+    @patch('os.makedirs', Mock())
+    @patch("pandas.DataFrame.to_csv")
+    @patch("pandas.read_csv")
+    def test_no_closure_type(self, mock_read, mock_csv):
```
[the remark this answers is not in the document]
Comment (jiayuanz3) [reply pyEpiabm/pyEpiabm/tests/test_func/test_closure_functional.py:134]: I don't think there exists such a function could be tested by .call_count. For instance, the scaling factor is multiplied in household_inf function, but no matter the place is closed or not, this function will be called anyway. The implementing logic is first to check if the intervention is active (here yes), then it assigns value to microcell's closure_start_time attribute (this step is done in __call__ in place_closure.py). So I'm not sure which function I can mock to test if place closure is present or not.

[the remark this answers is not in the document]
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/tests/test_func/test_closure_functional.py:135]: Ahh yes that makes sense! Maybe this test isn't needed then!

**Review by KCGallagher — APPROVED** [review]

**Comment by KCGallagher** [comment]
So the functional tests now take almost 2 minutes (as opposed to 5 seconds before). I think this is probably an unacceptable slowdown, so I want you to consider whether we can combine any of the tests, or if we need all of them - I don't think we need to test the impact of every intervention parameter for example. @rccreswell what are your thoughts on this?

**Comment by KCGallagher** [comment]
A gentler solution might be simply to reduce the population size in these tests - it maybe only needs to be about a hundred. Of course this may affect your answer to my previous [comment](https://github.com/SABS-R3-Epidemiology/epiabm/pull/191#discussion_r1128224289) about test stochasticity, but I would imagine this would be fine even on populations with n< 100.

**Pull request merged**